File: frontend/lib/rust_api/jobs/queue_stats.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../../config.dart';
import '../core.dart';

/// Q2 方案 B：`GET /api/v1/jobs/queue/stats`（须 header **`X-Toonflow-Internal-Token`** = 服务端 **`TOONFLOW_INTERNAL_OPS_TOKEN`**）。
Future<JobQueueStatsV1> fetchJobQueueStatsV1({
  required String internalOpsToken,
}) async {
  final uri = Uri.parse('$kApiBaseUrl/api/v1/jobs/queue/stats');
  final res = await http
      .get(uri, headers: {'x-toonflow-internal-token': internalOpsToken.trim()})
      .timeout(const Duration(seconds: 15));
  ensureHttpSuccess(res);
  final map = jsonDecode(res.body) as Map<String, dynamic>;
  return JobQueueStatsV1.fromJson(map);
}

class JobQueueStatsV1 {
  JobQueueStatsV1({
    required this.pending,
    required this.pendingClaimable,
    required this.running,
    required this.dead,
    required this.failedLast24h,
    required this.oldestClaimableQueuedAgeSecs,
    required this.pendingByKind,
  });

  final int pending;
  final int pendingClaimable;
  final int running;
  final int dead;
  final int failedLast24h;
  final int? oldestClaimableQueuedAgeSecs;
  final Map<String, dynamic> pendingByKind;

  factory JobQueueStatsV1.fromJson(Map<String, dynamic> json) {
    final rawKind = json['pending_by_kind'];
    final kindMap = rawKind is Map<String, dynamic>
        ? rawKind
        : <String, dynamic>{};
    return JobQueueStatsV1(
      pending: (json['pending'] as num?)?.toInt() ?? 0,
      pendingClaimable: (json['pending_claimable'] as num?)?.toInt() ?? 0,
      running: (json['running'] as num?)?.toInt() ?? 0,
      dead: (json['dead'] as num?)?.toInt() ?? 0,
      failedLast24h: (json['failed_last_24h'] as num?)?.toInt() ?? 0,
      oldestClaimableQueuedAgeSecs:
          (json['oldest_claimable_queued_age_secs'] as num?)?.toInt(),
      pendingByKind: kindMap,
    );
  }
}

File: backend/src/jobs/handlers/queue_stats.rs
//! Q2 方案 B：`GET /api/v1/jobs/queue/stats` — 只读 PG 队列聚合（与 worker **`job_queue_metrics`** 同源 **`QueueStats`**）。
//!
//! 门禁：环境变量 **`TOONFLOW_INTERNAL_OPS_TOKEN`**（非空）；请求须带 **`X-Toonflow-Internal-Token`** 与之完全一致。
//! 未配置 token 时返回 **403**，避免意外暴露队列深度。

use axum::{extract::State, http::HeaderMap, Json};
use serde::Serialize;
use serde_json::Value;
use utoipa::ToSchema;

use crate::error::ApiError;
use crate::jobs::queue::{PgQueue, Queue};
use crate::state::AppState;

use super::common::require_pool;

#[derive(Debug, Serialize, ToSchema)]
pub struct JobQueueStatsResponse {
    pub pending: i64,
    pub pending_claimable: i64,
    pub running: i64,
    pub dead: i64,
    pub failed_last_24h: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub oldest_claimable_queued_age_secs: Option<i64>,
    pub pending_by_kind: Value,
}

fn internal_ops_token_expected() -> Option<String> {
    std::env::var("TOONFLOW_INTERNAL_OPS_TOKEN")
        .ok()
        .map(|s| s.trim().to_owned())
        .filter(|s| !s.is_empty())
}

fn require_internal_ops_token(headers: &HeaderMap) -> Result<(), ApiError> {
    let Some(expected) = internal_ops_token_expected() else {
        return Err(ApiError::Forbidden(
            "job queue stats HTTP disabled (set TOONFLOW_INTERNAL_OPS_TOKEN)".into(),
        ));
    };
    let got = headers
        .get("x-toonflow-internal-token")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("");
    if got != expected.as_str() {
        return Err(ApiError::Unauthorized);
    }
    Ok(())
}

#[utoipa::path(
    get,
    path = "/api/v1/jobs/queue/stats",
    operation_id = "getJobQueueStatsV1",
    tag = "jobs",
    responses(
        (status = 200, description = "OK", body = JobQueueStatsResponse),
        (status = 401, description = "Unauthorized", body = crate::error::ErrorBody),
        (status = 403, description = "Forbidden", body = crate::error::ErrorBody),
        (status = 503, description = "Unavailable", body = crate::error::ErrorBody),
    )
)]
pub(crate) async fn get_job_queue_stats(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<Json<JobQueueStatsResponse>, ApiError> {
    require_internal_ops_token(&headers)?;
    let pool = require_pool(&state)?;
    let queue = PgQueue::new(pool.clone());
    let stats = queue
        .stats()
        .await
        .map_err(|e| ApiError::DatabaseError(e.to_string()))?;
    Ok(Json(JobQueueStatsResponse {
        pending: stats.pending,
        pending_claimable: stats.pending_claimable,
        running: stats.running,
        dead: stats.dead,
        failed_last_24h: stats.failed_last_24h,
        oldest_claimable_queued_age_secs: stats.oldest_claimable_queued_age_secs,
        pending_by_kind: stats.pending_by_kind_json,
    }))
}
